File: app/src/main/kotlin/com/kotlinextension/data/db/datasource/LocalUserDataSource.java
package com.kotlinextension.data.db.datasource;

import com.example.hassanusman.things2do.DataSource;

import java.util.List;

import io.reactivex.Flowable;

public class LocalUserDataSource implements DataSource<Task> {

    private TaskDao taskDao;
    public LocalUserDataSource(TaskDao taskDao){
        this.taskDao=taskDao;
    }

    @Override
    public Flowable<Task> get(Task task) {
        return taskDao.getTask(task.getTaskId());
    }

    @Override
    public Flowable<List<Task>> getAll() {
        return taskDao.getAllTasks();
    }

    @Override
    public void insert(Task task) {
        taskDao.insert(task);
    }

    @Override
    public void delete(Task task) {
        taskDao.deleteTask(task.getTaskId());
    }

    @Override
    public void update(Task task) {
        taskDao.updateTask(task.getTaskId(),task.getName(),task.getPending(),task.getDescription());
    }

}

File: app/src/main/kotlin/com/kotlinextension/data/db/datasource/LocalUserDataSource.kt
package com.kotlinextension.data.db.datasource

import com.kotlinextension.data.db.FailerApiResponse
import com.kotlinextension.data.db.dao.UsersDao
import com.kotlinextension.data.db.entity.User
import com.kotlinextension.data.remote.ApiService
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.schedulers.Schedulers

class LocalUserDataSource(private val userDao :UsersDao, private val apiService :ApiService) {
    fun get(bean :User) :Flowable<User> {
        return userDao.getUserById(bean.id)
    }

    fun getAll(error: FailerApiResponse) :Flowable<MutableList<User>> {
        val apiResult = apiService.getUsers(50)
        val dbResult = userDao.getAllUsers()

        return Flowable.combineLatest(apiResult, dbResult, )
        val stream = Flowable.create<Int>({subscriber ->
            apiService.getUsers(50)
                .subscribeOn(Schedulers.io())
                .subscribe({
                    it?.let {
                        userDao.insertUser(*it)
                    }
                }, {
                    error.onError(it)
                })
        }, BackpressureStrategy.MISSING)

        stream.subscribe(
            {`val` -> log.info("Subscriber received: {}", `val`)},
            {err -> log.error("Subscriber received error", err)}
        ) {log.info("Subscriber got Completed event")}

        apiService.getUsers(50)
            .subscribeOn(Schedulers.io())
            .subscribe({
                it?.let {
                   userDao.insertUser(*it)
                }
            }, {
                error.onError(it)
            })

        return userDao.getAllUsers()
    }

    fun insert(bean :User) {
        userDao.insertUser(bean)
    }

    fun delete(bean :User) {
        userDao.deleteUser(bean.id)
    }

    fun update(bean :User) {
        userDao.updateUser(bean)
    }
}
